vsintegration/src/FSharp.Editor/CodeFixes/ReplaceWithSuggestion.fs

```diff
@@ -13,54 +13,59 @@ open FSharp.Compiler.Diagnostics
-            let caretLine = sourceText.Lines.GetLineFromPosition(pos)
-            let fcsCaretLineNumber = Line.fromZ caretLinePos.Line
-            let lineText = caretLine.ToString()
-
-            let partialName =
-                QuickParse.GetPartialLongNameEx(lineText, caretLinePos.Character - 1)
-
-            let declInfo =
-                checkFileResults.GetDeclarationListInfo(Some parseFileResults, fcsCaretLineNumber, lineText, partialName)
-
-            let addNames (addToBuffer: string -> unit) =
-                for item in declInfo.Items do
-                    addToBuffer item.NameInList
-
-            for suggestion in CompilerDiagnostics.GetSuggestedNames addNames unresolvedIdentifierText do
-                let replacement = PrettyNaming.NormalizeIdentifierBackticks suggestion
-
-                do
-                    context.RegisterFsharpFix(
-                        CodeFix.ReplaceWithSuggestion,
-                        CompilerDiagnostics.GetErrorMessage(FSharpDiagnosticKind.ReplaceWithSuggestion suggestion),
-                        [| TextChange(context.Span, replacement) |]
-                    )
-        }
-        |> Async.Ignore
-        |> RoslynHelpers.StartAsyncUnitAsTask(context.CancellationToken)
+    override _.FixableDiagnosticIds = ImmutableArray.Create("FS0039", "FS0495")
+
+    override this.RegisterCodeFixesAsync context =
+        if context.Document.Project.IsFSharpCodeFixesSuggestNamesForErrorsEnabled then
+            context.RegisterFsharpFix this
+        else
+            Task.CompletedTask
+
+    interface IFSharpCodeFixProvider with
+        member _.GetCodeFixIfAppliesAsync context =
+            cancellableTask {
+                let! parseFileResults, checkFileResults =
+                    context.Document.GetFSharpParseAndCheckResultsAsync(nameof ReplaceWithSuggestionCodeFixProvider)
+
+                let! sourceText = context.GetSourceTextAsync()
+                let! unresolvedIdentifierText = context.GetSquigglyTextAsync()
+                let pos = context.Span.End
+                let caretLinePos = sourceText.Lines.GetLinePosition(pos)
+                let caretLine = sourceText.Lines.GetLineFromPosition(pos)
+                let fcsCaretLineNumber = Line.fromZ caretLinePos.Line
+                let lineText = caretLine.ToString()
+
+                let partialName =
+                    QuickParse.GetPartialLongNameEx(lineText, caretLinePos.Character - 1)
+
+                let declInfo =
+                    checkFileResults.GetDeclarationListInfo(Some parseFileResults, fcsCaretLineNumber, lineText, partialName)
+
+                let addNames addToBuffer =
+                    for item in declInfo.Items do
+                        addToBuffer item.NameInList
+
+                let suggestionOpt =
+                    CompilerDiagnostics.GetSuggestedNames addNames unresolvedIdentifierText
+                    |> Seq.tryHead
```

Comment:
This will be addressed in a followup (this currently doesn't work anyway - see the PR description). We need to change the framework to be able to produce multiple code fixes here - but it will be a messy change.